From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: athexweb3 <athexweb3@nirapod.org>
Date: Thu, 8 Oct 2026 19:09:38 +0600
Subject: [PATCH] safetylib: faster start-up with large global tables

Every byte of a global array is a variable, and the global arrays are
assigned at the start of the analysis. Adding the variables to the
environment one at a time, one lce per expression, and List.mem over
all the variables in SafetyCongr.change_environment were quadratic.
Reading one word of a global u64[N]: N = 4096, 17.9 s -> 1.4 s;
N = 8192, 50.8 s -> 3.1 s.

(cherry picked from commit 9fd4cc5e11fb5cf2613192e5b908a55f0f513c58)
---
 compiler/safetylib/domains/safetyCongr.ml |  7 +++-
 compiler/safetylib/domains/safetyNum.ml   | 49 ++++++++++++-----------
 2 files changed, 31 insertions(+), 25 deletions(-)

diff --git a/compiler/safetylib/domains/safetyCongr.ml b/compiler/safetylib/domains/safetyCongr.ml
index 4f2364362..5b6544e83 100644
--- a/compiler/safetylib/domains/safetyCongr.ml
+++ b/compiler/safetylib/domains/safetyCongr.ml
@@ -289,8 +289,11 @@ module AbsNumCongr : AbsNumType = struct
   let unify _ _ = assert false
 
   let change_environment t l =
-    (* Remove values not in [l] *)
-    let t = Mm.filter (fun k _ -> List.mem k l) t in
+    (* Remove values not in [l] (a hash table, as [l] can hold every byte
+       of the global arrays) *)
+    let h = Hashtbl.create (List.length l) in
+    List.iter (fun x -> Hashtbl.replace h x ()) l;
+    let t = Mm.filter (fun k _ -> Hashtbl.mem h k) t in
     List.fold_left (fun t v ->
         if not (Mm.mem v t)
         then Mm.add v Congr.top t
diff --git a/compiler/safetylib/domains/safetyNum.ml b/compiler/safetylib/domains/safetyNum.ml
index 463475b45..a26d08915 100644
--- a/compiler/safetylib/domains/safetyNum.ml
+++ b/compiler/safetylib/domains/safetyNum.ml
@@ -486,23 +486,30 @@ module AbsNumI (Manager : AprManager) (PW : ProgWrap) : AbsNumType = struct
       bound_texpr t (Mtexpr.var v)
     | _ -> Abstract1.bound_variable man t (avar_of_mvar v)
 
-  let env_add_mvar env v =
-    let add_single v env =
-      let av = avar_of_mvar v in
-      if Environment.mem_var env av then env
-      else
-        Environment.add env
-          (Array.of_list [av])
-          (Array.make 0 (Var.of_string "")) in
-
-    match v with
-    | Mglobal (AarraySlice _ )
-    | Mlocal  (AarraySlice _ ) ->
-      List.fold_left
-        (fun x y -> add_single y x) env
-        (u8_blast_var ~blast_arrays:true v)
-
-    | _ -> add_single v env
+  (* Add the variables [vs] missing from [env] (array slices are blasted
+     into bytes). They are added at once: [Environment.add] sorts the whole
+     environment. *)
+  let env_add_mvars env vs =
+    let blast v = match v with
+      | Mglobal (AarraySlice _ )
+      | Mlocal  (AarraySlice _ ) -> u8_blast_var ~blast_arrays:true v
+      | _ -> [v] in
+    let avs =
+      List.concat_map blast vs
+      |> List.rev_map avar_of_mvar
+      |> List.filter (fun av -> not (Environment.mem_var env av))
+      |> List.sort_uniq Var.compare in
+    if avs = [] then env
+    else Environment.add env (Array.of_list avs) (Array.make 0 (Var.of_string ""))
+
+  (* [prepare_env] for a list of expressions, with a single [lce]. *)
+  let prepare_env_list env mexprs =
+    let vars =
+      List.concat_map Mtexpr.get_var mexprs
+      |> List.rev_map avar_of_mvar
+      |> List.sort_uniq Var.compare
+      |> Array.of_list in
+    Environment.lce env (Environment.make vars (Array.make 0 (Var.of_string "")))
 
   let e_complex e =
     (is_relational ()) && (Mtexpr.contains_mod e)
@@ -532,14 +539,10 @@ module AbsNumI (Manager : AprManager) (PW : ProgWrap) : AbsNumType = struct
     let v_weaks = List.rev v_weaks in
     
     (* If v_copies are not in the environment, we add them *)
-    let env = List.fold_left (fun env v_cp ->
-        env_add_mvar env v_cp
-      ) (Abstract1.env a) v_copies in 
+    let env = env_add_mvars (Abstract1.env a) v_copies in
 
     (* We add the variables in the expressions to the environment *)
-    let env = List.fold_left (fun env e ->
-        prepare_env env e
-      ) env es in
+    let env = prepare_env_list env es in
     let a = Abstract1.change_environment man a env false in
 
     (* If the domain is relational, and if e contains a modulo, then we just
